src/Data/OffHeap.hs

```diff
@@ -0,0 +1,117 @@
+-- ensure predictable latency (/e.g./ in distributed applications). The cost is
+-- that memory management is much more explicit: the programmer has to allocate
+-- and free memory manually. Safety (in particular that every pointer is freed)
+-- is enforced by linear types, which constrain usage, in particular
+-- sharing. Off-heap data types also have less convenient syntax since they are
+-- not directly supported by the compiler.
+--
+-- This module focuses on /pure/ off-heap data. That is data types like standard
+-- Haskell. The only difference is that their lifetime is not managed by the
+-- GC. Despite calling @malloc@ and @free@ under the hood, the entire API is
+-- pure, and does not make calls in IO.
+--
+-- The allocation API is organised around a notion of memory 'Pool'. From the API
+-- point of view, a pool serves as a source of linearity. That is: it ensures
+-- that the allocation primitive need not take a continuation to delimit its
+-- lifetime. Besides convenience, it avoids needlessly preventing functions from
+-- being tail-recursive.
+--
+-- Pools play another role: resilience to exceptions. If an exception is raised,
+-- all the data in the pool is deallocated. It does not, however, impose a stack
+-- discipline: data in pools is normally freed by the destruction primitives of ,
+-- only in case of exception are the pool deallocated in a stack-like
+-- manner. Moreover, pool A can have data pointing to pool B, while at the same
+-- time, pool B having data pointing to pool A.
+--
+-- Functions in this module are meant to be qualified.
+
+-- TODO: add link to an example in module header
+-- TODO: change some words into link to the relevant API entry in the above description.
+
+module Data.OffHeap
+  ( Pool
+  , withPool
+  , Box
+  , alloc
+  , deconstruct
+  ) where
+
+import Foreign.Marshal.Alloc
+import Foreign.Ptr
+import Foreign.Storable
+import Prelude.Linear
+import System.IO.Unsafe
+import qualified Unsafe.Linear as Unsafe
+
+-- TODO: ignoring exceptions for the moment. So that I can get some tests to
+-- work first.
+
+-- TODO: Briefly explain the Dupable-reader style of API, below, and fix
+-- details.
+
+-- | A 'Pool' can be 'consume'-ed. This is a no-op: it does not deallocate the
+-- data in that pool. It cannot as there may still be accessible data in the
+-- pool. It simply makes it impossible to add new data to the pool. It is
+-- actually necessary to so consume a pool allocated with 'withPool' in order to
+-- write a well-typed scope @Pool ->. Unrestricted b@.
+data Pool = Pool
+
+-- TODO: document individual functions
+
+withPool :: (Pool ->. Unrestricted b) ->. Unrestricted b
+withPool scope = scope Pool
+
+instance Consumable Pool where
+  consume Pool = ()
+
+instance Dupable Pool where
+  dup Pool = (Pool, Pool)
+
+-- XXX: this indirection is possibly not necessary. It's here because the inner
+-- Ptr must be unrestricted (in order to implement deconstruct at the moment).
+data Box a where
```

Comment:
Why the non-descript `Box` name rather than just using `Unrestricted` here? Just because `Box a` shorter than `Unrestricted (Ptr a)`? Or as an abstraction barrier (hide the `Ptr`)?

Reply:
It's an abstraction barrier around `Ptr`. It differs from `Ptr` in that it is pure. Also it will hold some more data for exception tolerance.